packages/db/src/resources/projects/resolveContractInstances.ts

```diff
@@ -0,0 +1,214 @@
+
+    contractInstances.push(...stepContractInstances);
+  }
+
+  return contractInstances;
+}
+
+/**
+ * Steps backwards through name record history for a given set of name records,
+ * yielding an array of past resources in a breadth-first search manner.
+ *
+ * At each step, yields a [somewhat HACKy] `skip` function, to omit specific
+ * indexes from further consideration.
+ *
+ * Returns when all name records histories have been exhausted.
+ */
+async function* findResourcesHistories<N extends NamedCollectionName>(options: {
+  collectionName: N;
+  nameRecords: (SavedInput<"nameRecords"> | undefined)[];
+  workspace: Workspace;
+}): AsyncIterable<
+  {
+    [K in "skip" | N]: "skip" extends K
+      ? (...indexes: number[]) => void
+      : (IdObject<N> | undefined)[];
+  }
+> {
+  const { collectionName, workspace } = options;
+  let { nameRecords } = options;
+
+  do {
+    const skip = (...indexes: (number | undefined)[]) => {
+      for (const index of indexes) {
+        if (typeof index === "number") {
+          nameRecords[index] = undefined;
+        }
+      }
+    };
+
+    // @ts-ignore
+    yield {
+      skip,
+      [collectionName]: nameRecords.map(nameRecord =>
+        nameRecord && nameRecord.resource
+          ? ({ id: nameRecord.resource.id } as IdObject<N>)
+          : undefined
+      )
+    };
+
+    // preserving order, iterate to next set of previous records
+    nameRecords = await workspace.find(
+      "nameRecords",
+      nameRecords.map(nameRecord =>
+        nameRecord && nameRecord.previous ? nameRecord.previous : undefined
+      )
+    );
+  } while (nameRecords.find(nameRecord => nameRecord));
+}
+
+/**
+ * Given a list of candidate networks, returns a subset list that are each
+ * ancestor to the network currently known to a project as a given name.
+ */
+async function filterProjectNetworkAncestors(options: {
+  project: IdObject<"projects">;
+  network: DataModel.ResourceNameInput;
+  candidates: IdObject<"networks">[];
+  workspace: Workspace;
+  info: graphql.GraphQLResolveInfo;
+}): Promise<IdObject<"networks">[]> {
+  const { project, network, workspace, info } = options;
+
+  // short-circuit early to avoid extra queries and to avoid index lookup guard
+  if (options.candidates.length === 0) {
+    return [];
+  }
+
+  const candidates = await workspace.find("networks", options.candidates);
+
+  // find earliest among candidates to specify minimumHeight to ancestors query
+  const earliestCandidate = candidates
+    .slice(1)
+    .reduce(
+      (earliest, network) =>
+        earliest.historicBlock.height < network.historicBlock.height
+          ? earliest
+          : network,
+      candidates[0]
+    );
+
+  // query ancestors for project network name
+  const { network: { ancestors = [] } = {} } = await delegateToSchema({
+    schema: info.schema,
+    operation: "query",
+    fieldName: "project",
+    returnType: info.schema.getType("Project") as graphql.GraphQLOutputType,
```

Comment:
A little confused about this -- why do you need to cast this type?

Reply:
Cause graphql's types aren't friendly, it seems

Reply:
(The reason is that `getType` doesn't know if it's returning an output type or an input type, but `returnType` demands only output types. But _we_ know that `"Project"` will be an output type)